### PR TITLE
Cycle detection feat

### PR DESCRIPTION
**Pull request by g2hchang**
Proposed completion for cycle detection in the graph feature for issue #55 .
I created a cycle detect button on top right corner, and when the user clicks it, it will go through the graph to find all the vertices that are part of cycle and highlight it as red.

**Example of cycle :** 
screen shot 2017-09-16 at 6 42 18 pm

screen shot 2017-09-16 at 6 37 02 pm

**Example of no cycle :** 
screen shot 2017-09-16 at 6 42 02 pm

screen shot 2017-09-16 at 6 37 16 pm

### CODE REVIEW

**Comment by amgrubb** [comment]
How does the user clear the red highlighting after a cycle is detected?

**Comment by g2hchang** [comment]
To clear the red highlight, the cycle in the graph needs to be resolved.

**Comment by amgrubb** [comment]
Does the user needs to re-run cycle detection to clear the red highlighting on the nodes once the cycle has been removed? or 
When a link is deleted then the highlighting is removed?

**Comment by g2hchang** [comment]
Added an option for user to clear highlight by adding an option in clear button tab.
So, user has two options to clear : 
1) Break the cycle in the graph and re-run the cycle checker
2) Use the clear button to clear all the highlight

Workflow : 

**1. User checks for cycle and is detected**
screen shot 2017-09-17 at 10 17 20 pm

**2. User clicks clear all cycle highlight under clear tab**
screen shot 2017-09-17 at 10 17 28 pm

**Now red highlight is all gone.**
screen shot 2017-09-17 at 10 17 41 pm

**Comment by amgrubb** [comment]
I really like the details you provided in this pull request and the screenshots. I now understand how it is supposed to work.

This is a good opportunity to create a brief wiki page explaining how cycle detection works. You can use the same screenshots.

File: leaf-ui/index.html [hunk 1]
```diff
@@ -17,6 +17,9 @@
         <link rel="stylesheet" type="text/css" href="./css/style.css" />
         <link rel="stylesheet" type="text/css" href="./css/analysis_style.css" />
 
+        <!--zebra css -->
```
Comment (amgrubb) [inline leaf-ui/index.html]: Add additional comments why zebra css is added.
For example:

File: leaf-ui/index.html [hunk 2]
```diff
@@ -73,6 +77,8 @@
 
             <button id="symbolic-btn" class="blue-btn btn">Model Constraints</button>
             <button id="analysis-btn" class="green-btn btn">Analysis</button>
+            <button id="cycledetect-btn" class="pink-btn btn">Cycle Checker</button>
```
Comment (amgrubb) [inline leaf-ui/index.html]: Sorry, I should have specified this.
Can we change the name of this button to "Detect Cycles"

File: leaf-ui/index.html [hunk 3]
```diff
@@ -130,6 +136,10 @@
 
         <script src="./scripts/main.js"></script>
 
+        <!--Zebrajs -->
```
Comment (amgrubb) [inline leaf-ui/index.html]: Again include more comments why we use Zebra and what it is.

File: leaf-ui/scripts/main.js [hunk 4]
```diff
@@ -262,6 +261,127 @@ $('#model-init-btn').on('click', function(){
...
+		}
+	}
+	var elements = graph.getElements();
+
+	js_object = null;
+	jslinks = null;
+})
+
+//Cycle-deteciton algorithm
+function cycleCheck(links, verticies){
```
Comment (amgrubb) [inline leaf-ui/scripts/main.js]: Need to include a comment: 

> Based on the Detect Cycle in a Directed Graph Algorithm discussed at 
http://www.geeksforgeeks.org/detect-cycle-in-a-graph/

**Review by amgrubb — CHANGES_REQUESTED** [review]
Great job on the cycle detection!
See inline comments.
Also, please move the "node_modules" folder into the "plugins" folder, and update links accordingly.

File: leaf-ui/index.html (hunk 1 above)
Comment (amgrubb) [reply leaf-ui/index.html]: Zebra is used for creating Visually Appealing Dialog boxes?

**Comment by g2hchang** [comment]
Hi Alicia, I have added comments and removed the node_modules directory.

**Review by amgrubb — APPROVED** [review]

**Review by amgrubb — CHANGES_REQUESTED** [review]
Missed one change, please rename the button.

Update: I figured out I can change it through the web interface.

File: leaf-ui/package-lock.json [hunk 5]
```diff
@@ -0,0 +1,16 @@
+{
```
Comment (amgrubb) [inline leaf-ui/package-lock.json]: What is the file for and do you still need "zebra_dialog" portion of it. It looks like you removed the rest of the zebra dialog files?

Comment (amgrubb) [reply leaf-ui/package-lock.json]: Please comment the file.

Comment (amgrubb) [reply leaf-ui/package-lock.json]: Also, should this file live in the plugins folder?

**Review by amgrubb — CHANGES_REQUESTED** [review]
Sorry, I have a couple more questions.

File: leaf-ui/package-lock.json (hunk 5 above)
Comment (g2hchang) [reply leaf-ui/package-lock.json]: Sorry, I removed the zebra_dialog. That was another library for customizing dialog, but I chose to use sweetalert instead. And package.json is not needed, so I will remove them and push it again.

**Comment by g2hchang** [comment]
I removed the package.json. I am using sweetalert.min.js for dialog customization and I have commented about the library.

**Review by amgrubb — APPROVED** [review]

**Comment by amgrubb** [comment]
Excellent Job, now we need to test this on `develop`.

**Pull request merged**